### PR TITLE
[MOT] add c++ implementation for flow_statistics

### PR DESCRIPTION
**Pull request by FlyingQianMM**
At each frame, count the following data from the the beginning to current frame:
* the number of identifiers 
* the number of identifiers across the horizontal center line in a top to down manner
* the number of identifiers across the horizontal center line in a down to top manner

At a manual-set interval (unit: seconds), count the number of identifiers from the the beginning to end frame.

### CODE REVIEW

File: deploy/pptracking/include/pipeline.h [hunk 1]
```diff
@@ -81,11 +83,16 @@ class Pipeline {
   // Run pipeline in stream
   void RunMOTStream(const cv::Mat img,
                     const int frame_id,
+                    const int video_fps,
+                    const Rect entrance,
                     cv::Mat out_img,
                     std::vector<std::string>* records,
-                    std::vector<int>* count_list,
+                    std::set<int>* count_set,
```
Comment (jerrywgz) [inline deploy/pptracking/include/pipeline.h:90]: count_set->id_set 更好点？

File: deploy/pptracking/src/main.cc [hunk 2]
```diff
@@ -54,7 +54,8 @@ DEFINE_bool(trt_calib_mode,
             "If the model is produced by TRT offline quantitative calibration, "
             "trt_calib_mode need to set True");
 DEFINE_bool(tiny_obj, false, "Whether tracking tiny object");
-DEFINE_bool(count, false, "Whether counting after tracking");
+DEFINE_bool(count, false, "Whether counting in and out numbers after tracking");
```
Comment (jerrywgz) [inline deploy/pptracking/src/main.cc]: entrance_count更好些？

File: deploy/pptracking/include/pipeline.h (hunk 1 above)
Comment (FlyingQianMM) [reply deploy/pptracking/include/pipeline.h:90]: Done. Thanks.

File: deploy/pptracking/src/main.cc (hunk 2 above)
Comment (FlyingQianMM) [reply deploy/pptracking/src/main.cc]: Done. Thanks.

**Review by jerrywgz — DISMISSED** [review]

**Review by jerrywgz — DISMISSED** [review]

File: deploy/pptracking/include/postprocess.h [hunk 3]
```diff
@@ -40,9 +41,17 @@ cv::Mat VisualizeTrackResult(const cv::Mat& img,
 // Pedestrian/Vehicle Counting
 void FlowStatistic(const MOTResult& results,
                    const int frame_id,
-                   std::vector<int>* count_list,
-                   std::vector<int>* in_count_list,
-                   std::vector<int>* out_count_list);
+                   const int secs_interval,
+                   const bool count,
```
Comment (qingqing01) [inline deploy/pptracking/include/postprocess.h]: count -> is_counting

File: deploy/pptracking/src/pipeline.cc [hunk 4]
```diff
@@ -221,6 +228,18 @@ void Pipeline::PredictMOT(const std::string& video_path) {
 
     fclose(fp);
     LOG(INFO) << "txt result output saved as " << result_output_path.c_str();
+
+    result_output_path = output_dir_ + OS_PATH_SEP + "flow_statistic.txt";
```
Comment (qingqing01) [inline deploy/pptracking/src/pipeline.cc:232]: LOG(INFO)应该放在232行下面吧

File: deploy/pptracking/src/pipeline.cc [hunk 5]
```diff
@@ -221,6 +228,18 @@ void Pipeline::PredictMOT(const std::string& video_path) {
...
+    if ((fp = fopen(result_output_path.c_str(), "w+")) == NULL) {
+      printf("Open %s error.\n", result_output_path);
```
Comment (qingqing01) [inline deploy/pptracking/src/pipeline.cc:234]: 这里是直接报错吗？

File: deploy/pptracking/src/postprocess.cc [hunk 6]
```diff
@@ -101,10 +102,80 @@ cv::Mat VisualizeTrackResult(const cv::Mat& img,
...
+  if (frame_id % video_fps == 0 && frame_id / video_fps % secs_interval == 0) {
+    curr_interval_count = interval_id_set->size();
+    os << ", Count during " << secs_interval
+       << " secs: " << curr_interval_count;
+    interval_id_set->clear();
+  }
+  os << "\n";
+  std::string record = os.str();
+  records->push_back(record);
+  std::cout << record;
```
Comment (qingqing01) [inline deploy/pptracking/src/postprocess.cc]: std:cout -> LOG(INFO)

File: deploy/pptracking/include/postprocess.h (hunk 3 above)
Comment (FlyingQianMM) [reply deploy/pptracking/include/postprocess.h]: 已修改为do_entrance_counting。感谢！

File: deploy/pptracking/src/pipeline.cc (hunk 4 above)
Comment (FlyingQianMM) [reply deploy/pptracking/src/pipeline.cc:232]: 这个打印的是上一个保存文件

File: deploy/pptracking/src/pipeline.cc (hunk 5 above)
Comment (FlyingQianMM) [reply deploy/pptracking/src/pipeline.cc:234]: output_dir在进入该函数之前会创建，所以这里提示无法打开并返回就好

File: deploy/pptracking/src/postprocess.cc (hunk 6 above)
Comment (FlyingQianMM) [reply deploy/pptracking/src/postprocess.cc]: 已修改。感谢！

**Review by jerrywgz — DISMISSED** [review]

**Review by nemonameless — DISMISSED** [review]

**Review by jerrywgz — APPROVED** [review]

**Review by nemonameless — APPROVED** [review]

**Pull request merged**